src/dynimage.rs

```diff
@@ -639,17 +639,19 @@ impl GenericImage for DynamicImage {
 }
 
 /// Decodes an image and stores it into a dynamic image
-fn decoder_to_image<'a, I: ImageDecoder<'a>>(codec: I) -> ImageResult<DynamicImage> {
-    let color = codec.color_type();
-    let (w, h) = codec.dimensions();
-    let buf = codec.read_image()?;
+fn decoder_to_image<'a, I: ImageDecoder<'a>>(decoder: I) -> ImageResult<DynamicImage> {
+    let color_type = decoder.color_type();
+    let (w, h) = decoder.dimensions();
+    let total_bytes = decoder.total_bytes();
 
-    // TODO: Avoid this cast by having ImageBuffer use u64's
-    assert!(w <= u64::from(u32::max_value()));
-    assert!(h <= u64::from(u32::max_value()));
-    let (w, h) = (w as u32, h as u32);
+    if total_bytes > usize::max_value() as u64 {
+        return Err(ImageError::InsufficientMemory);
+    }
+
+    let mut buf = vec![0; total_bytes as usize];
```

Comment:
Maybe add a note that both zero-extension and cutoff lead to a useful comparison and make the reverse conversion correct. A better *limit* here might also be `isize::max_value()` which is the maximum number of bytes any `Vec` will try to allocate on 32-bit platforms. But maybe this should not act like a limit and use fallible conversion instead:

```
let total_bytes = usize::try_from(total_bytes)
    .map_err(|_| ImageError::InsufficientMemory)?;
```

Reply:
I thought about this a bit more, and I think the right thing to do in these cases is actually to panic. If total_bytes / scanline_bytes is under isize::max_value() but still more than the size of RAM, then the user is going to get a OOM panic. It seems silly to try to protect them if it is even larger than that.